src/groups/heisenberg_group.jl

```diff
@@ -0,0 +1,200 @@
+
+@doc raw"""
+    HeisenbergGroup{T} <: AbstractDecoratorManifold{ℝ}
+
+Heisenberg group `HeisenbergGroup(n)` is the group of ``(n+2)×(n+2)`` matrices [BinzPods:2008](@cite)
+
+```math
+\begin{bmatrix} 1 & \mathbf{a} & c \\
+\mathbf{0} & I_n & \mathbf{b} \\
+0 & \mathbf{0} & 1 \end{bmatrix}
+```
+
+where ``I_n`` is the ``n×n`` unit matrix, ``\mathbf{a}`` is a row vector of length ``n``,
+``\mathbf{b}`` is a column vector of length ``n`` and ``c`` is a real number.
+The group operation is matrix multiplication.
+
+The left-invariant metric on the manifold is used.
+"""
+const HeisenbergGroup{T} = LieGroup{
+    ℝ,MatrixMultiplicationGroupOperation,Manifolds.HeisenbergMatrices{T}
+}
+
+function HeisenbergGroup(n::Int; parameter::Symbol=:type)
+    Hm = Manifolds.HeisenbergMatrices(n; parameter=parameter)
+    return HeisenbergGroup{typeof(Hm).parameters...}(
+        Hm, MatrixMultiplicationGroupOperation()
+    )
+end
+
+function _heisenberg_a_view(M::HeisenbergGroup, p)
+    n = ManifoldsBase.get_parameter(M.manifold.size)[1]
+    return view(p, 1, 2:(n + 1))
+end
+function _heisenberg_b_view(M::HeisenbergGroup, p)
+    n = ManifoldsBase.get_parameter(M.manifold.size)[1]
+    return view(p, 2:(n + 1), n + 2)
+end
+
+@doc raw"""
+    exp(M::HeisenbergGroup, ::Identity{MatrixMultiplicationGroupOperation}, X)
+
+Lie group exponential for the [`HeisenbergGroup`](@ref) `M` of the vector `X`.
+The formula reads
+```math
+\exp\left(\begin{bmatrix} 0 & \mathbf{a} & c \\
+\mathbf{0} & 0_n & \mathbf{b} \\
+0 & \mathbf{0} & 0 \end{bmatrix}\right) = \begin{bmatrix} 1 & \mathbf{a} & c + \mathbf{a}⋅\mathbf{b}/2 \\
+\mathbf{0} & I_n & \mathbf{b} \\
+0 & \mathbf{0} & 1 \end{bmatrix}
+```
+where ``I_n`` is the ``n×n`` identity matrix, ``0_n`` is the ``n×n`` zero matrix
+and ``\mathbf{a}⋅\mathbf{b}`` is dot product of vectors.
+"""
+function Base.exp(M::HeisenbergGroup, e::Identity{MatrixMultiplicationGroupOperation}, X)
+    q = similar(X)
+    exp!(M, q, e, X)
+    return q
+end
+
+function ManifoldsBase.exp!(
+    M::HeisenbergGroup, q, ::Identity{MatrixMultiplicationGroupOperation}, X
+)
+    n = ManifoldsBase.get_parameter(M.manifold.size)[1]
+    copyto!(q, I)
+    a_view = _heisenberg_a_view(M, X)
+    b_view = _heisenberg_b_view(M, X)
+    q[1, 2:(n + 1)] .= a_view
+    q[2:(n + 1), n + 2] .= b_view
+    q[1, n + 2] = X[1, n + 2] + dot(a_view, b_view) / 2
+    return q
+end
+
+@doc raw"""
+    exp(M::HeisenbergGroup, p, X)
```

Comment:
Can we maybe use G (not M) for the group, and g,h (not p,q) ? That avoids name clashes when referring to the (underlying) manifold

(sorry this should have been included in the review before, which GitHub decided to leave empty, and delete what I already typed here before. no clue why)

Reply:
Sure, I've updated the naming.